### PR TITLE
DEVEXP-527: Starting on tests for cascading collections and permissions.

### PR DESCRIPTION
**Pull request by BillFarber**
Is this what you're thinking for the tests?

Note the "setCascadeCollections" method I added. I don't know if I'm going the right direction for that. I just needed something to make the test.

### CODE REVIEW

**Comment by BillFarber** [comment]
The new version of this PR is intended simply to share an initial implementation strategy and to discuss it's pros and cons.

File: src/test/java/com/marklogic/client/ext/file/CascadeCollectionsAndPermissionsFileProcessorTest.java
Comment (rjrudin) [inline src/test/java/com/marklogic/client/ext/file/CascadeCollectionsAndPermissionsFileProcessorTest.java:1]: Just do "CascadeCollectionsAndPermissionsTest" so the test class name is about the use case as opposed to the classes involved. The existing tests for collections/permissions unfortunately follow the "(class name)Test" pattern - I try to only use that for pure unit tests that focus on several methods of a test. Ignore my lousy naming with those classes, "(Use case)Test" is the way to go.

File: src/test/java/com/marklogic/client/ext/file/CascadeCollectionsAndPermissionsFileProcessorTest.java [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+		assertTrue(optionalTestFile.isPresent());
+		testFile = optionalTestFile.get();
+		assertTrue(testFile.getDocumentMetadata().getCollections().contains(CHILD_COLLECTION));
+		assertFalse(testFile.getDocumentMetadata().getCollections().contains(PARENT_COLLECTION));
+		assertTrue(testFile.getDocumentMetadata().getPermissions().get("rest-reader").contains(DocumentMetadataHandle.Capability.READ));
+		assertNull(testFile.getDocumentMetadata().getPermissions().get("rest-writer"));
+	}
+
+	@Test
+	public void ChildHasSettingsUsedWhenParentHasNoSettings() {
```
Comment (rjrudin) [inline src/test/java/com/marklogic/client/ext/file/CascadeCollectionsAndPermissionsFileProcessorTest.java:68]: Still do lowerCamel for these test names, just like the first one.

File: src/test/java/com/marklogic/client/ext/file/CascadeCollectionsAndPermissionsFileProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		modulesClient.newServerEval().xquery("cts:uris((), (), cts:true-query()) ! xdmp:document-delete(.)").eval();
+	}
+
+	@Test
+	public void parentSettingsOnlyCascadeWhenChildHasNoSettings() {
+		String directory = "src/test/resources/process-files/cascading-metadata-test/parent1-withCP";
+		GenericFileLoader loader = new GenericFileLoader(client);
+		List<DocumentFile> files = loader.loadFiles(directory);
+
+		Optional<DocumentFile> optionalTestFile = files.stream().filter(file -> "/child1_1-noCP/test.json".equals(file.getUri())).findFirst();
```
Comment (rjrudin) [inline src/test/java/com/marklogic/client/ext/file/CascadeCollectionsAndPermissionsFileProcessorTest.java:50]: It's okay to test the DocumentFile objects that are returned, but what we really care about is that the modules were loaded correctly - users care about that, ml-gradle users will never see the List. Use the `verifyCollections` and `verifyPermissions` methods I added - they'll make the test much shorter and easier to read. If you want, include assertions on the List that's returned as well, though we can safely assume that if the modules were loaded correctly, then the List is going to be correct as well (I would not bother with those assertions).

**Review by rjrudin — CHANGES_REQUESTED** [review]
Looks good, just a couple test changes.

File: src/main/java/com/marklogic/client/ext/file/CascadingPropertiesDrivenDocumentFileProcessor.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Properties;
+import java.util.Stack;
+
+/**
+ * Adds a stack to store Properties objects while traversing a directory tree.
+ */
+public abstract class CascadingPropertiesDrivenDocumentFileProcessor extends PropertiesDrivenDocumentFileProcessor {
```
Comment (rjrudin) [inline src/main/java/com/marklogic/client/ext/file/CascadingPropertiesDrivenDocumentFileProcessor.java:27]: Remove "public" here - another mistake I made was making everything public initially.

**Review by rjrudin — CHANGES_REQUESTED** [review]
Looks good, just one tiny change and then merge away.

**Pull request merged**